pkg/operator/observe_config.go

```diff
@@ -18,54 +18,68 @@ import (
 
+	configinformers "github.com/openshift/client-go/config/informers/externalversions"
+	configlistersv1 "github.com/openshift/client-go/config/listers/config/v1"
 	operatorconfigclientv1alpha1 "github.com/openshift/cluster-kube-apiserver-operator/pkg/generated/clientset/versioned/typed/kubeapiserver/v1alpha1"
 	kubeapiserveroperatorinformers "github.com/openshift/cluster-kube-apiserver-operator/pkg/generated/informers/externalversions"
 )
 
-type observeConfigFunc func(kubernetes.Interface, *rest.Config, map[string]interface{}) (map[string]interface{}, error)
+type Listers struct {
+	imageConfigLister configlistersv1.ImageLister
+	endpointLister    corelistersv1.EndpointsLister
+	configmapLister   corelistersv1.ConfigMapLister
+}
+
+type observeConfigFunc func(Listers, map[string]interface{}) (map[string]interface{}, error)
 
 type ConfigObserver struct {
 	operatorConfigClient operatorconfigclientv1alpha1.KubeapiserverV1alpha1Interface
 
-	kubeClient kubernetes.Interface
-
 	// queue only ever has one item, but it has nice error handling backoff/retry semantics
 	queue workqueue.RateLimitingInterface
 
 	rateLimiter flowcontrol.RateLimiter
 	// observers are used to build the observed configuration. They are called in
 	// order and are expected to mutate the config based on cluster state.
 	observers []observeConfigFunc
+
+	// listers are used to by config observers to retrieve necessary resources
+	listers Listers
 }
 
 func NewConfigObserver(
 	operatorConfigInformers kubeapiserveroperatorinformers.SharedInformerFactory,
 	kubeInformersForKubeSystemNamespace kubeinformers.SharedInformerFactory,
+	configInformer configinformers.SharedInformerFactory,
 	operatorConfigClient operatorconfigclientv1alpha1.KubeapiserverV1alpha1Interface,
-	kubeClient kubernetes.Interface,
 ) *ConfigObserver {
 	c := &ConfigObserver{
 		operatorConfigClient: operatorConfigClient,
-		kubeClient:           kubeClient,
 
 		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ConfigObserver"),
 
 		rateLimiter: flowcontrol.NewTokenBucketRateLimiter(0.05 /*3 per minute*/, 4),
 		observers: []observeConfigFunc{
 			observeEtcdEndpoints,
 			observeClusterConfig,
+			observeInternalRegistryHostname,
+		},
+		listers: Listers{
+			imageConfigLister: configInformer.Config().V1().Images().Lister(),
+			endpointLister:    kubeInformersForKubeSystemNamespace.Core().V1().Endpoints().Lister(),
+			configmapLister:   kubeInformersForKubeSystemNamespace.Core().V1().ConfigMaps().Lister(),
 		},
 	}
 
 	operatorConfigInformers.Kubeapiserver().V1alpha1().KubeAPIServerOperatorConfigs().Informer().AddEventHandler(c.eventHandler())
 	kubeInformersForKubeSystemNamespace.Core().V1().Endpoints().Informer().AddEventHandler(c.eventHandler())
 	kubeInformersForKubeSystemNamespace.Core().V1().ConfigMaps().Informer().AddEventHandler(c.eventHandler())
+	configInformer.Config().V1().Images().Informer().AddEventHandler(c.eventHandler())
```

Comment:
here's a neat side effect.  If you don't wait for your informers to sync once, you can race and destroy your config.